### DISCUSSION TITLE
Serialize base models as python primitives

### DISCUSSION THREAD

**Discussion by AnghileriDavide** [pydantic/pydantic#3996] [Feature Request]
**It would be nice to have the possibility to define models that can be serialized to python primitives like strings.**
Pydantic assumes that the JSON representation of an object is something in the form of a mapping, so you don't really get to serialize an object as a primitive type (yet). 

An example of why such a feature could be useful is the case in which we have a `MySlug` base model that represent a versioning of a service, which could be a complex python object (containing for example the service name and version) but is then passed around between services requests and responses as an encoded string.

E.g.

```python
from pydantic import BaseModel

class MySlug(BaseModel):

    protocol_version: int 
    content: str
    ...
    
    @classmethod
    def from_slug(cls, slug: str) -> "MySlug":
        """
        Parses the `slug`, constructing the `Slug` internal representation.
        Validates that the slug protocol matches the `protocol` class attribute.
        """
        pass
    
     def to_slug(self) -> str:
        """
        Returns the slugified representation of the `Slug`.
        """
        pass
        
     @classmethod
     def __get_validators__(cls):
        # my custom validators
```

For simplicity the content of the slug is a `str` in the example, but it could be a more complex python object containing various information useful to properly encode the versioning of the system.
Then when we can use it inside a `MyResponse` base model, e.g:

```python
class MyResponse(BaseModel):
    slug: MySlug
    ...
```

Supposing that the `MySlug.content` get encoded into the slug `"v1~service:1.0"` by the `to_slug()` method, it would be nice if when the `MyResponse` get serialized or encoded as a dict we get:
`{'slug': 'v1~service:1.0'}`
instead of 
`{'slug': {'content': ...}}`.

---

We already tried the following approaches:
- Defining the `__root__` attribute but it cannot be used with other fields.
- Creating a `SlugStr(str)` type with an attribute `slug: Type[Slug]` that is a string but returns the Slug in the `__get_validators__` method. However we didn't like that we have to create the various `SlugStr` dinamically with a Factory in order to define the `MyResponse.slug` as a Union of multiple `SlugStr(SlugV1)`, `SlugStr(SlugV2)`, ...

**Comment by hbock-42** [comment]
We are very interested about this too
